### ISSUE TITLE
Azure Active Directory do not refresh auth token

### ISSUE THREAD

**Issue by jeliasson** [redwoodjs/redwood#1672]
Authentication provider `azureActiveDirectory` does not refresh token after expiry, which is `3600 seconds`. As of canary `0.23.1-canary.10` the user is redirected to the authentication page, but it will need some more work on the provider to renew the token.

**Related Issues**
#1636 [Tracking] Validate token refresh issue across providers

**Related PR**
#1675 Azure Active Directory refresh access token

/cc @dac09

**Comment by jeliasson** [comment]
@dac09 @thedavidprice Assign this issue to me. Thanks!

**Comment by dthyresson** [comment]
> @dac09 @thedavidprice Assign this issue to me. Thanks!

Assigned to @jeliasson.

**Comment by jeliasson** [comment]
Fixed and merged in cab9e5f. Closing issue.

@andrewlamyw
Feel free to test this from canary build (`yarn rw upgrade -t canary`) and let us know if you have any issues. 

/cc @dac09

**Comment by andrew-hwahin** [comment]
> @andrewlamyw
> Feel free to test this from canary build (`yarn rw upgrade -t canary`) and let us know if you have any issues.

@jeliasson  Thanks for your effort. 🙏 I'm facing this error after upgrading to the canary version 0.23.1-canary.23

![image](https://user-images.githubusercontent.com/61768800/106227843-07da6f80-6225-11eb-859e-4615a6d79abe.png)
![image](https://user-images.githubusercontent.com/61768800/106227909-26d90180-6225-11eb-846b-8b7da0107dfa.png)

**Comment by Tobbe** [comment]
@andrew-hwahin There is a breaking change in canary. We've changed an import and component name. Full details will be in the release notes for the next release. But basically what you have to do the same changes as I've done here:

https://github.com/redwoodjs/redwood/pull/1639/files#diff-478b4309175ff97877735ec43c9c1305479332e22b178c06d737f6c66ff841a1

![image](https://user-images.githubusercontent.com/30793/106248536-6d246580-6211-11eb-8f4b-8ae143551cc7.png)

**Comment by jeliasson** [comment]
@Tobbe Thanks for the clarification.
@andrew-hwahin Below is my setup.

### `web/src/index.js`
```js
import ReactDOM from 'react-dom'
import { azureActiveDirectoryClient } from 'src/lib/microsoft/azure/msal/client'

import { AuthProvider } from '@redwoodjs/auth'
import { RedwoodApolloProvider } from '@redwoodjs/web/apollo'
import { FatalErrorBoundary } from '@redwoodjs/web'

import FatalErrorPage from 'src/pages/FatalErrorPage'

import Routes from 'src/Routes'

ReactDOM.render(
  <FatalErrorBoundary page={FatalErrorPage}>
    <AuthProvider
      client={azureActiveDirectoryClient}
      type="azureActiveDirectory"
    >
      <RedwoodApolloProvider>
        <Routes />
      </RedwoodApolloProvider>
    </AuthProvider>
  </FatalErrorBoundary>,
  document.getElementById('root')
)
```
### `web/src/lib/microsoft/azure/msal/client.js`
```js
import { UserAgentApplication } from 'msal'

export const azureActiveDirectoryConfig = {
  auth: {
    clientId: process.env.AZURE_ACTIVE_DIRECTORY_CLIENT_ID,
    authority: process.env.AZURE_ACTIVE_DIRECTORY_AUTHORITY,
    redirectUri: process.env.AZURE_ACTIVE_DIRECTORY_REDIRECT_URI,
    postLogoutRedirectUri:
      process.env.AZURE_ACTIVE_DIRECTORY_LOGOUT_REDIRECT_URI,
  },
}

export const azureActiveDirectoryClient = new UserAgentApplication(
  azureActiveDirectoryConfig
)
```

Notice that the new `getToken()` method expects environment variable `AZURE_ACTIVE_DIRECTORY_CLIENT_ID` to be set. If you are using another environment variable, I suggest you pass `options` to `getToken(options?)`, e.g.

```js
const token = getToken({
   scopes: [process.env.ANOTHER_ENVIRONMENT_VARIABLE],
})
```
See https://github.com/redwoodjs/redwood/blob/main/packages/auth/src/authClients/azureActiveDirectory.ts#L20

I'll create a PR in `redwoodjs.com` to reflect this.

**Comment by andrew-hwahin** [comment]
A million thanks for the guidance and efforts 🙏 @Tobbe @jeliasson 

I'm able to upgrade to the canary version successfully and so far I'm using it with no problem, I will update here again if I encountered any.😁

**Comment by jeliasson** [comment]
@andrew-hwahin You are welcome. Be noted that https://github.com/redwoodjs/redwood/pull/1703 was merged a few hours ago and this will most likely result in that [new docs](https://deploy-preview-534--redwoodjs.netlify.app/docs/authentication#azure-active-directory) being published as part of Redwood 0.24.0 release.

While this won't make any breaking changes, it clarifies how `logIn` and `getToken` behaves. Please let us know if there is something unclear in the docs. 🙏
